### PR TITLE
Formats-bsd: unit tests performance

### PR DESCRIPTION
**Pull request by sbesson**
See https://trello.com/c/ISuLct1l/125-formats-bsd-unit-tests-take-too-long-to-run

This PR proposes to reduce the time spent in the default Formats BSD unit tests with the main bottleneck being the `SixteenBitLosslessJPEG2000Test` class which loops over every short value from -32,768 to +32,767.

#### Test classes

- add an increment field in `SixteenBitLosslessJPEG2000Test` controlling the loop of `testLosslessPixels`
- set this increment to `1024` by default
- define another test class and use the TestNG Factory annotation to set the increment to `1` and test all short values

#### Build infrastructure

- leave `testng.xml` consumed by the default Maven `test` target unchanged
- define a new `testng-long-running.xml` configuration file and `test-long-running` Ant target allowing to run the full test using the entire Short range.

### Testing instructions
- check Travis is green with this PR included and the time spent in the Maven matrix component is reduced and/or check that `mvn test` completes faster
- check that `ant test` runs the `SixteenBitLosslessJPEG2000LongRunningTest` and tests all short values

### CODE REVIEW

File: components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000LongRunningTest.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Long-running test case which outlines the problems seen in    
+ * trac.openmicroscopy.org/ome/ticket/6278.
+ */
+public class SixteenBitLosslessJPEG2000LongRunningTest {
+
+  @Factory
+  public Object[] factoryMethod() {
+    return new Object[] { new SixteenBitLosslessJPEG2000Test(1)};
```
Comment (mtbc) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000LongRunningTest.java]: spacing

File: components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java [hunk 2]
```diff
@@ -66,12 +66,21 @@
   private static final Logger LOGGER =
     LoggerFactory.getLogger(SixteenBitLosslessJPEG2000Test.class);
 
+  private int increment = 1024;
+
+  public SixteenBitLosslessJPEG2000Test() {
+  }
+
+  public SixteenBitLosslessJPEG2000Test(int increment) {
+      this.increment=increment;
```
Comment (mtbc) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java]: spaces around =

File: components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java [hunk 3]
```diff
@@ -66,12 +66,21 @@
   private static final Logger LOGGER =
     LoggerFactory.getLogger(SixteenBitLosslessJPEG2000Test.class);
 
+  private int increment = 1024;
+
+  public SixteenBitLosslessJPEG2000Test() {
+  }
```
Comment (mtbc) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java]: a more idiomatic pattern might be to have `private final int increment;` and in this constructor do `this(1024);` but this approach is fine too

**Comment by mtbc** [comment]
`ant test`: Without this PR I have 943 Formats-BSD unit tests running in 3m34 and with this PR I have 943 tests running in 3m56.
`mvn test`: Without this PR it runs in 3m37 and with this PR it runs in 2m31 where for both `SixteenBitLosslessJPEG2000Test` is green in `formats-bsd/target/surefire-reports/`.

The above seem typical over multiple runs. Slightly slower `ant test` seems worth rather faster `mvn test`. This is looking good.

**Comment by sbesson** [comment]
Thanks for the testing @mtbc. Your last metrics raises the question of whether the 1min increase for the default Maven unit tests is significant enough or if we need to spend more time identifying the bottlenecks /cc @simleo

File: components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java (hunk 3 above)
Comment (simleo) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java]: You mean `this(increment)`?

Comment (mtbc) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java]: No, as how the no-arg constructor could specify 1024 default. (Fewer state changes = easier reasoning.)

Comment (mtbc) [inline components/formats-bsd/test/loci/formats/utests/SixteenBitLosslessJPEG2000Test.java]: (The pattern's perhaps a bit like `doChange` or `newUserInGroup` in `AbstractServerTest` where we chain the methods to add arguments to cover for our lack of default values in Java, yet another thing the language could have learned from Modula-3.)

**Comment by simleo** [comment]
On my laptop, the running time for the whole Maven test suite is 3:49 without and 2:19 with the patch. The BSD section goes down to 1:05 from 2:37, so it's more than twice as fast. The BSD suite is still the longest running at 1:05, but `SixteenBitLosslessJPEG2000Test` was the main bottleneck and this PR removes it. We should merge, then we can look for the next big offender.

**Comment by sbesson** [comment]
@mtbc @simleo: pushed

**Comment by mtbc** [comment]
Lovely, thank you. :+1:

**Comment by rleigh-dundee** [comment]
This looks fine and good to merge.

The only question I have WRT travis is whether we want to be able to selectively enable these long running tests for a given matrix combination e.g. jdk7.  If ant is going to be dropped in the future, then I think it would be nice to have the option to run the long running tests with maven.  Maybe this is already possible?

**Comment by sbesson** [comment]
Thanks, merging this. Re https://github.com/openmicroscopy/bioformats/pull/2446#issuecomment-225135922, one of the major scoping tasks associated with the Ant -> Maven transition proposal is certainly to review and list all the tasks which are specific to one build system.

**Pull request merged**